Virtual Memory SV39 Tests

Zain2050 opened the pull request:
<FOR DOC UPDATES FILL ONLY DESCRIPTION AND RELATED ISSUES SECTION AND REMOVE THE OTHERS>

## Description

> Tests have been added to verify the functionality of the `SV39` virtual memory extension. `Core-V Wally` has been used as the `dut` for these tests.

### Related Issues

> NA

### Ratified/Unratified Extensions

- [x] Ratified
- [ ] Unratified

### List Extensions

> SV39

### Reference Model Used

- [x] SAIL
- [ ] Spike
- [ ] Other - < SPECIFY HERE >

### Mandatory Checklist:

  - [x] All tests are compliant with the test-format spec present in this repo ?
  - [x] Ran the new tests on RISCOF with SAIL/Spike as reference model successfully ?
  - [ ] Ran the new tests on RISCOF in [coverage mode](https://riscof.readthedocs.io/en/stable/commands.html#coverage)
  - [ ] Link to Google-Drive folder containing the new coverage reports ([See this](https://github.com/riscv-non-isa/riscv-arch-test/blob/main/CONTRIBUTION.md#uploading-test-stats) for more info): < SPECIFY HERE >

### Optional Checklist:

  - [x] Were the tests hand-written/modified ?
  - [x] Have you run these on any hard DUT model ? Please specify name and provide link if possible in the description
  - [x] If you have modified arch_test.h Please provide a detailed description of the changes in the Description section above.

UmerShahidengr committed:
Spliting dev & sv39 (#2)

UmerShahidengr committed:
Removed all sv32 updates from this branch

Zain2050 committed:
SV39 tests

Zain2050 committed:
Fixed indentation in arch_test.h

sudo-apt-abdullah committed:
RV64VM_vm_permissions_cg = 84.84%

sudo-apt-abdullah committed:
vm_add_feature_cg = 100%

hamza-1821 committed:
vm_mprv_S_mode.S

hamza-1821 committed:
mstatus_tvm_test.S

hamza-1821 committed:
vm_mprv_U_mode.s

sudo-apt-abdullah committed:
vm_permission = 92.00%

sudo-apt-abdullah committed:
Merge branch 'sv39' of github.com:Zain2050/riscv-arch-test into sv39

Zain2050 committed:
mstatus at 100

Zain2050 committed:
Modified template to reduce executable size

Zain2050 committed:
Added DAU tests

Zain2050 committed:
Merge branch 'dev' of https://github.com/riscv-non-isa/riscv-arch-test into sv39

Zain2050 committed:
Merged separate levels into one file

Zain2050 committed:
Simplified Code, Reviewed it & Added Proper Comments

Zain2050 committed:
Merge branch 'dev' of https://github.com/riscv-non-isa/riscv-arch-test into sv39

Zain2050 committed:
Removed DAU Fault Tests

UmerShahidengr committed:
Merge branch 'dev' into sv39

Zain2050 committed:
Renamed some files

Zain2050 committed:
pull new changes from dev to sv39

Zain2050 committed:
Resolved merge conflict in testmacros.h

Zain2050 committed:
Minor correction in MXR tests

allenjbaum commented:
That's OK, someone should be documenting it - preferably the author....

Zain2050 committed:
Improved test desc in svnapot & svpbmt tests

UmerShahidengr committed:
Merge branch 'dev' into sv39

allenjbaum reviewed (changes requested):
on line 1422, don't use a constant, use a variable name, e.g. CAUSE_FETCH_ACCESS in this case (0xC might be wrong in this case; that is CAUSE_FETCH_PAGE_FAULT

allenjbaum commented on riscv-test-suite/env/test_macros.h:
```diff
@@ -232,17 +232,33 @@ Mend_PMP:                                    ;\
+    or _PAR, _PAR, _PR                  /* Combine PPN with permissions */                  ;\
+    .if (level==2)                      /* Level 2 (1GB superpage) */                       ;\
+        LA(_TR1, rvtest_Sroot_pg_tbl)   /* Load root page table address */                  ;\
+        LI(_TR0, ((VA>>30)&0x1FF)<<3)   /* Calculate index for LEVEL2 (bits 38:30) */       ;\
```
We should have a field extract macro e.g. 
   EXTRACT(_VAL_,_POS_,_BITWIDTH)  ((_VAL_>>_POS_)&((1<<_BITWIDTH_)-1))
and define RV64_LVL0_PPNPOS, RV64_LVL1_PPNPOS, RV64_LVL2_PPNPOS, RV64PPNWIDTH)
It could probably be used in a lot of places. I think we've defined XLEN_WIDTH somewhere.

Zain2050 replied:
Yes, this is a good idea. I'll add it.

allenjbaum commented on riscv-test-suite/env/test_macros.h:
```diff
@@ -232,17 +232,33 @@ Mend_PMP:                                    ;\
+// in a single line.
+// args: PA_LBL: Label of Physical Address, PERMS: permissions in hex
+// args: VA: Virtual Address in hex, level: Level to store at (0, 1, or 2)
+#define PTE_SETUP_RV39_New(PA_LBL, PERMS, VA, level)                                         \
```
IS using a fixed reg# for PA_LBL and  and Perms a problem? The old macro had the register parameters expicit in the parameter definitions instead of explicit in the code .  Could it instead reuse TR0 or TR1, or have them supplied as explicit parameters to the new macro?

Zain2050 replied:
In the previous macro we actually have to store physical address and premissions in registers and then pass those registers. This updated macro sets up a PTE in one line. 
It has been changed to match the PTE_SETUP_SV32 macro which is already present. TR0 and TR1 can't be reused as they store page table address and offset. 
Same as the SV32 tests, we have made sure that a0 and a1 (fixed registers) are not being used anywhere else.
But if you want, I can change the macro to accept registers as parameters.

allenjbaum replied:
Once _PAR is calculated (first 3 lines), then _PR is dead, and can be used instead of one of the temp registers (BUT only assuming that that _PR Xreg  will not be reused outside the macro!, as opposed to be reloaded) 
You should add comments as to which of the registers must be preserved by the macro somewhere if you make that change

allenjbaum replied:
The optimization  of changing code to reduce the number of registers just isn't important enough to waste time on. But: replacing a0 and a1 with passed in register numbers: I still want to see that done.

allenjbaum reviewed (changes requested):
in 1422, use defined variable CAUSE_FETCH_ACCESS instead of 0xC (which is incorrect, I think, because that is CAUSE_FETCH_PAGE_FAULT instead of 0x1

Zain2050 committed:
Updated arch test

Zain2050 commented:
I have changed 0xC to CAUSE_FETCH_PAGE_FAULT. 
Well, we actually need to check for **Fetch page fault** instead of **Fetch access fault**, as this fault occurs when we are trying to execute pages with X or V unset. I think the comment caused this confusion. I have improved the comment as well.

allenjbaum commented on riscv-test-suite/rv64i_m/vm_sv39/src/vm_canonical_S_mode.S:
```diff
@@ -0,0 +1,251 @@
+.section .text.init
+.globl rvtest_entry_point
+rvtest_entry_point:
+RVMODEL_BOOT												// This test supports max 255 words for RVMODEL_BOOT
```
good comments, but they're scolled off my screen for no good reason., which makes it harder to review. Could you move them in closer?

allenjbaum replied:
We will do this in the next iteration

UmerShahidengr committed:
Merge branch 'dev' into sv39

UmerShahidengr committed:
Update arch_test.h

Added the CAUSE_FETCH_ACCESS check within CAUSE_FETCH_PAGE_FAULT to handle the infinite trap loop case

Signed-off-by: Umer Shahid <umer.shahid@10xengineers.ai>

allenjbaum reviewed (approved):
This isn't perfect, but it works, and has been checked
